### PR TITLE
[barcode-scanner] Deprecate and add warning

### PR DESCRIPTION
**Pull request by alanjhughes**
# Why
Closes ENG-10921
We are deprecating barcode scanner and recommending to use `expo-camera` instead.

# How
Add deprecation docs and warning.

### CODE REVIEW

File: packages/expo-barcode-scanner/src/BarCodeScanner.tsx [hunk 1]
```diff
@@ -15,6 +15,8 @@ const { BarCodeType, Type } = ExpoBarCodeScannerModule;
 
 const EVENT_THROTTLE_MS = 500;
 
+let warnAboutDeprecation = false;
```
Comment (tsapeta) [inline packages/expo-barcode-scanner/src/BarCodeScanner.tsx:18]: Wouldn't `warnedAboutDeprecation` fit better? 😅 
```suggestion
let warnedAboutDeprecation = false;
```

File: packages/expo-barcode-scanner/src/BarCodeScanner.tsx [hunk 2]
```diff
@@ -138,6 +144,15 @@ export class BarCodeScanner extends React.Component<BarCodeScannerProps> {
     barCodeTypes: Object.values(BarCodeType),
   };
 
+  componentDidMount(): void {
+    if (!warnAboutDeprecation) {
```
Comment (tsapeta) [inline packages/expo-barcode-scanner/src/BarCodeScanner.tsx:148]:
```suggestion
    if (!warnedAboutDeprecation) {
```

File: packages/expo-barcode-scanner/src/BarCodeScanner.tsx [hunk 3]
```diff
@@ -138,6 +144,15 @@ export class BarCodeScanner extends React.Component<BarCodeScannerProps> {
...
+      console.warn(
+        'BarCodeScanner has been deprecated and will be removed in a future SDK version. Please use `expo-camera` instead'
+      );
+      warnAboutDeprecation = true;
```
Comment (tsapeta) [inline packages/expo-barcode-scanner/src/BarCodeScanner.tsx:152]:
```suggestion
      warnedAboutDeprecation = true;
```

**Review by tsapeta — APPROVED** [review]

File: packages/expo-barcode-scanner/src/BarCodeScanner.tsx (hunk 1 above)
Comment (alanjhughes) [reply packages/expo-barcode-scanner/src/BarCodeScanner.tsx:18]: I was just about to change it 😅

**Review by lukmccall — APPROVED** [review]

File: packages/expo-barcode-scanner/src/BarCodeScanner.tsx [hunk 4]
```diff
@@ -138,6 +144,15 @@ export class BarCodeScanner extends React.Component<BarCodeScannerProps> {
     barCodeTypes: Object.values(BarCodeType),
   };
 
+  componentDidMount(): void {
+    if (!warnedAboutDeprecation) {
+      console.warn(
+        'BarCodeScanner has been deprecated and will be removed in a future SDK version. Please use `expo-camera` instead'
```
Comment (brentvatne) [inline packages/expo-barcode-scanner/src/BarCodeScanner.tsx:150]: we should have an fyi page that explains how to migrate from barcodescanner to expo-camera, can you write one up and add a link to it in this warning @alanjhughes? to create an fyi page, just add a markdown file to https://github.com/expo/fyi

Comment (alanjhughes) [reply packages/expo-barcode-scanner/src/BarCodeScanner.tsx:150]: No problem 👍

**Comment by alanjhughes** [comment]
@Bardiamist - can you provide a reproducible example of the issue?

**Comment by Bardiamist** [comment]
I see, need to migrate to `expo-camera/next`.
Is it stable and recommended for replace or better wait for now and keep deprecated `expo-barcode-scanner` until `expo-camera` will be replaced by `expo-camera/next` fully?

**Comment by alanjhughes** [comment]
@Bardiamist - It is stable, we will keep it as `next` for this SDK cycle and if you have any problems with it, you can report them and fall back to the the old version while we fix it. If all goes well, we will fully deprecate the old version in SDK 51.

**Pull request merged**